### PR TITLE
[IDE] Overall energy consumption and UI hang improvements

### PR DESCRIPTION
**Pull request by Therzok**
There are various small things added up.

Most notably, the assembly browser is now more responsive and faster to load, and should no longer leak.

Other small improvements

### CODE REVIEW

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserViewContent.cs [hunk 1]
```diff
@@ -169,40 +176,29 @@ public void FindDerivedClasses ()
...
-						LoggingService.LogError ("Error while adding project " + project.Name + " to the tree.", e);
-					}
-				}
-				await Task.WhenAll (allTasks).ContinueWith (delegate {
-					Runtime.RunInMainThread (delegate {
+				var token = cts.Token;
+
+				var workspace = IdeApp.TypeSystemService.GetWorkspaceAsync (IdeApp.ProjectOperations.CurrentSelectedSolution)
+					.ContinueWith (t => {
+						if (token.IsCancellationRequested)
```
Comment (mrward) [inline main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserViewContent.cs]: Was wondering if this needed to handle the task exception but it looks like GetWorkspaceAsync would not cause an exception.

**Review by mrward — APPROVED** [review]
👍

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserViewContent.cs (hunk 1 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserViewContent.cs]: I had another small patch around this, but it's just a fake hang improvement.

**Review by sevoku — APPROVED** [review]

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs [hunk 2]
```diff
@@ -914,24 +914,26 @@ void HPaneExpose (object sender, Gtk.ExposeEventArgs args)
...
-					try {
-						if (currentAssembly != null) {
-							OpenFromAssembly (url, currentAssembly);
-						} else {
-							OpenFromAssemblyNames (url);
-						}
-					} catch (Exception e) {
-						LoggingService.LogError ("Error while opening the assembly browser with id:" + url, e);
+				if (d.IsFaulted) {
+					LoggingService.LogError ("Failed to load assemblies", d.Exception);
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs]: Shouldn't be a return here ?

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs [hunk 3]
```diff
@@ -98,33 +98,32 @@ static string GetSource (ITypeDefinition type)
...
 		{
 			var type = (ITypeDefinition)dataObject;
-			var list = new System.Collections.ArrayList ();
 			if (type.DirectBaseTypes.Any ())
-				list.Add (new BaseTypeFolder (type));
+				builder.AddChild (new BaseTypeFolder (type));
 			bool publicOnly = Widget.PublicApiOnly;
 
-			foreach (var field in type.Fields.OrderBy (m => m.Name, StringComparer.InvariantCulture)) {
+			foreach (var field in type.Fields) {
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs]: Shouldn't the names be ordered ?

**Review by mkrueger — APPROVED** [review]

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs (hunk 2 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs]: definitions will be empty if it's faulted, so I guess I can do a return there too.

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs (hunk 3 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs]: Yes, but the sorting is done at treenode level. We don't need to do it here.

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs (hunk 2 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs]: I'm changing a few things here, to better handle faulting.

Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs]: I've thought this through.

There's a few scenarios which are handled by explicitly not returning on faulted.

a) IsFaulted is set when even just one task is faulted, but after all have run. What if we can actually find the type we're looking for? 
b) If the current assembly loader task fails, it'll be caught by the try/catch a bit further below.

Added a comment.

File: main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs (hunk 3 above)
Comment (Therzok) [reply main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs]: See the [comparer](https://github.com/mono/monodevelop/blob/5bbe3317491f6f588f5fb7938ed1c1ef8f1169f2/main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/AssemblyBrowserTypeNodeBuilder.cs#L60)

**Comment by Therzok** [comment]
@slluis PTAL

**Comment by Therzok** [comment]
Rebased and fixed conflicts.

**Review by slluis — APPROVED** [review]

**Pull request merged**